source/cuda/cuda_solver.cu

```diff
@@ -228,8 +228,7 @@ CudaSolver<VectorType>::CudaSolver(
 
     // We need to cast away the const because we need to change the format
     // of _matrix to the format supported by amgx
-    auto &amgx_matrix = const_cast<SparseMatrixDevice<value_type> &>(
-        *(cuda_operator->get_matrix()));
+    auto &amgx_matrix = *(cuda_operator->get_matrix());
```

Comment:
This is really confusing. In the end, `matrix` is changed heavily here. For me, it would make more sense if the constructor copies the object.

Reply:
I don't understand what's the problem here.

Reply:
Passing `op` as `std::shared_ptr` to a `const` object feels conceptually weird if we change the wrapped object.

Reply:
I see but 
  1) I will not copy the object. This is the most expensive part memory-wise of the algorithm. 
  2) We do have a few variables (include the data in `SparseMatrixDevice`) that are `mutable` and while we shouldn't use `mutable` this is out of the scope of this PR.

Reply:
Let's document these problems and move on.